### PR TITLE
Automatic buildpack additions

### LINKED ISSUE #387: Missing automatic buildpack detection
Some applications need additional buildpacks to run correctly in Cloud Foundry. For example the Lamp-App needs some PHP-Extensions (mysql, mysqli).
I think a good start is to implement the logic for PHP. Afterwards we hopefully can expand easily with more combinations. 
- [x] maybe there is a library which could help us. That must be investigated
- [x] recognize the used language
- [x] detect the combination of the language and the used services (PHP & Mysql-Database)
- [x] insert the right Extensions in the right way
- [x] create tests to ensure correct behaviour

### PR DESCRIPTION
**Pull request by JensMueller2709**
resolves #387

### CODE REVIEW

File: server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryFileCreator.java [hunk 1]
```diff
@@ -177,22 +174,12 @@ private boolean addMatchedServices(List<ServiceOffering> services,
...
-        buildPackAdditionsJson.put(BUILDPACK_OBJECT_PHP, buildPacks);
-        String path;
-        if (app.getPathToApplication() != null) {
-            path = String.format("%s/%s", app.getPathToApplication(), BUILDPACK_FILEPATH_PHP);
-        } else {
-            path = BUILDPACK_FILEPATH_PHP;
-        }
-        fileAccess.access(path).append(buildPackAdditionsJson.toString(4)).close();
+        CloudFoundryBuildpackDetection buildpackDetection = new CloudFoundryBuildpackDetection(app, fileAccess);
+        buildpackDetection.detectBuildpackAdditions();
```
Comment (nfode) [inline server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryFileCreator.java]: How do you know if the Buildpack addition was successfull? There is no return value neither a exception is thrown if something gs wrong.

File: server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpack.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package org.opentosca.toscana.plugins.cloudfoundry.application.buildpacks;
+
+/**
+ contains default buildpack values
+ */
+public class CloudFoundryBuildpack {
+
+    //will be expanded
+    public static final String[] DEFAULT_PHP_BUILDPACKS = new String[]{"bz2", "zlib", "curl", "mcrypt"};
```
[the remark this answers is not in the document]
Comment (JensMueller2709) [reply server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpack.java]: yes, when adding manually buildpacks the default values would be overwritten. That's why I'm adding them here.

File: server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpackDetection.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+                //TODO: expand with more languages
+            }
+        }
+    }
+
+    private void addBuildpackAdditonsPHP() throws JSONException, IOException {
+
+        if (application.getServices().containsValue(CloudFoundryServiceType.MYSQL)) {
+            JSONObject buildPackAdditionsJson = new JSONObject();
```
[the remark this answers is not in the document]
Comment (JensMueller2709) [reply server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpackDetection.java]: Yes I take this on my refactoring list. I will do it in a separate PR.

File: server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpackDetection.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            buildPackAdditionsJson.put(BUILDPACK_OBJECT_PHP, buildPacks);
+            String path;
+            if (application.getPathToApplication() != null) {
+                path = String.format("%s/%s", application.getPathToApplication(), BUILDPACK_FILEPATH_PHP);
+            } else {
+                path = BUILDPACK_FILEPATH_PHP;
+            }
+            fileAccess.access(path).append(buildPackAdditionsJson.toString(4)).close();
```
[the remark this answers is not in the document]
Comment (JensMueller2709) [reply server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/application/buildpacks/CloudFoundryBuildpackDetection.java]: I use it for better readability

File: server/src/integration/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryLampIT.java [hunk 5]
```diff
@@ -36,7 +36,7 @@
     private final String envSpace = System.getenv(CF_ENVIRONMENT_SPACE);
 
     public CloudFoundryLampIT() {
-        super(new CloudFoundryPlugin());
+        super(new Plugin());
```
Comment (c-mueller) [inline server/src/integration/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryLampIT.java]: I forgot to mention that you should have excluded the CloudFoundry Plugin and lifecylce class from this suggestion. As well as the corresponting tests. Sorry for that "inconvenience"

File: server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryFileCreator.java (hunk 1 above)
Comment (JensMueller2709) [reply server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryFileCreator.java]: I check the addition in the BuildpackDetector itself (in line 42 - 46). 
Is it sufficient or do you prefer another way to do it?

Comment (nfode) [reply server/src/main/java/org/opentosca/toscana/plugins/cloudfoundry/CloudFoundryFileCreator.java]: The problem is you just log if something goes wrong and the transformation does not abort.

**Comment by c-mueller** [comment]
Could you inform me on the status of this PR?

**Comment by JensMueller2709** [comment]
I think we could merge the PR. I already did all the requested changes from you and @nfode

**Pull request merged**